src/sentry/relay/config/metric_extraction.py

```diff
@@ -265,13 +281,77 @@ def _convert_widget_query_to_metric(
     return metrics_specs
 
 
+def _is_widget_query_low_cardinality(widget_query: DashboardWidgetQuery, project: Project):
+    """
+    Checks cardinality of existing widget queries before allowing the metric spec, so that
+    group by clauses with high-cardinality tags are not added to the on_demand metric.
+
+    New queries will be checked upon creation and not allowed at that time.
+    """
+    params: dict[str, Any] = {
+        "statsPeriod": "1d",
+        "project_id": project.id,
+        "project_objects": [project],
+        "organization_id": project.organization_id,  # Organization id has to be specified to not violate allocation policy.
+    }
+    params["statsPeriod"] = "1d"
+    start, end = get_date_range_from_params(params)
+    params["start"] = start
+    params["end"] = end
+
+    query_killswitch = options.get("on_demand.max_widget_cardinality.killswitch")
+    if query_killswitch:
+        return False
+
+    if len(widget_query.columns) == 0:
+        # No columns means no high-cardinality tags.
+        return True
+
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+    cache_key = f"check-widget-query-cardinality:{widget_query.id}"
+    cardinality_allowed = cache.get(cache_key)
+
+    if cardinality_allowed is not None:
+        return cardinality_allowed
+
+    unique_columns = [f"count_unique({column})" for column in widget_query.columns]
+
+    query_builder = QueryBuilder(
+        dataset=Dataset.Discover,
+        params=params,
+        selected_columns=unique_columns,
+        config=QueryBuilderConfig(
+            transform_alias_to_input_format=True,
+        ),
+    )
+
+    results = query_builder.run_query(Referrer.METRIC_EXTRACTION_CARDINALITY_CHECK.value)
+    processed_results = query_builder.process_results(results)
+
+    try:
+        for index, column in enumerate(widget_query.columns):
+            count = processed_results["data"][0][unique_columns[index]]
+            if count > max_cardinality_allowed:
+                cache.set(cache_key, False)
+                raise HighCardinalityWidgetException(
+                    f"Cardinality exceeded for dashboard_widget_query:{widget_query.id} with count:{count} and column:{column}"
+                )
+    except HighCardinalityWidgetException as error:
+        sentry_sdk.capture_exception(error)
+        return False
+
+    cache.set(cache_key, True)
```

Comment:
It'll be the default here, which is low. Thinking about this, not only do we need to set it high, we need to add some jitter to avoid thundering herd once we have more and more of these since they are making discover queries.

Reply:
Set it to 1d + 25% jitter. We probably want it even higher in practice for GA, but this should suffice for LA.

Reply:
I see timeout being set on the other cache.set() calls; had you intended to add it here as well?